nemo/collections/asr/parts/k2/utils.py

```diff
@@ -0,0 +1,205 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Copyright (c) 2020, Xiaomi CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import itertools
+import os
+from pickle import UnpicklingError
+from typing import List, Optional, Tuple, Union
+
+import k2
+import torch
+
+from nemo.utils import logging
+
+
+def create_supervision(input_lengths: torch.Tensor,) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Creates a special supervisions tensor from input lengths.
+    These supervisions are required for some k2 methods.
+    """
+    supervisions = torch.stack(
+        (torch.tensor(range(input_lengths.shape[0])), torch.zeros(input_lengths.shape[0]), input_lengths.cpu(),), 1,
+    ).to(torch.int32)
+    # the duration column has to be sorted in decreasing order
+    order = torch.argsort(supervisions[:, -1], descending=True)
+    return supervisions[order], order
+
+
+def invert_permutation(indices: torch.Tensor) -> torch.Tensor:
+    """Produces a tensor of reverse permutation for a given indices.
+    
+    Copied from https://github.com/k2-fsa/snowfall/blob/master/snowfall/common.py
+    """
+    ans = torch.zeros(indices.shape, device=indices.device, dtype=torch.long)
+    ans[indices] = torch.arange(0, indices.shape[0], device=indices.device)
+    return ans
+
+
+def make_blank_first(
+    blank_idx: int, log_probs: torch.Tensor, targets: torch.Tensor
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Puts blank logits at the first place in input log_probs tensor.
+    """
+    index = list(range(log_probs.shape[-1]))
+    del index[blank_idx]
+    index = torch.tensor([blank_idx] + index).to(log_probs.device)
+    new_log_probs = torch.index_select(log_probs, -1, index)
+    # TODO (alaptev): replace targets + 1 with torch.where to work for non-last blank_id
+    return new_log_probs, None if targets is None else targets + 1
+
+
+def load_graph(graph_path: str) -> k2.Fsa:
```

Comment:
Return type 'k2.Fsa'

Reply:
Done in [7b2b4a4](https://github.com/NVIDIA/NeMo/pull/3351/commits/7b2b4a4ba4e34848d526d8c4ac53903bfbfe2fd2)